site-src/geps/gep-1426.md

```diff
@@ -146,6 +141,23 @@ This should be enforced by controller validation, which should set an `Accepted:
 
 The logic described here is a bit verbose, but could be validated with conformance tests, and is intended to allow an `HTTPRoute` configuring mesh traffic rules to additionally specify a `Gateway` `parentRef` (or `HTTPRoute` `parentRef` if [GEP-1058: Route Inclusion and Delegation](https://github.com/kubernetes-sigs/gateway-api/pull/1085) moves forward) to apply the same routing rules when exposing a mesh service outside the cluster.
 
+### Namespace boundaries
+
+In a mesh, routes can be configured by two personas:
+
+* Service producers, who want to modify behavior of *inbound* requests to their services
+* Service consumers, who want to modify behavior of *outbound* requests to other services.
+
+While these concepts are not directly exposed in the API, a route is implicitly fulfilling one of these rolls and behaves differently depending on the role.
+
+A route is a producer route when the `parentRef` refers to a service in the *same namespace*. This route SHOULD apply to all incoming requests to the service, including from clients in other namespaces.
+
+Note: Some implementations may only be able to apply routes on client-side proxies. As a result, these will likely only apply to requests from clients who are also in the mesh.
+
+A route is a consumer route when the `parentRef` refers to a service in *another namespace*. Unlike producer routes, consumer routes are scoped only the same namespace. This ensures that for traffic between two namespaces, another unrelated namespace cannot modify their traffic.
```

Comment:
I changed the title; I think they are interconnected enough its a bit hard to split out.

Reply:
I get what these consumer routes is trying to achieve, but I think that there's one big thing that needs further explanation (which Rob raised above, but I think is most relevant to this bit):

What happens when a consumer route's backendRef to another namespace targets a producer Service in that namespace that is itself a parentRef to a Route?

To put this another way, does consumer-side traffic management override producer-side traffic-management? That seems like a very fraught idea, as it would make it so that consumer routes could end up preferring old versions etc, without the producer knowing about it.

If it doesn't, is it that a producer Service is not a valid target for the backendRef of a consumer Service? If so, how do you know that it's a consumer Service? There's nothing on the producer Service to indicate it, and it seems likely that the case would arrive that a consumer rule targets a Service that _later_ becomes the parentRef of a Route, which would mean that it would become invalid _when the owner of a producer service made a change_, which also seems... not ideal.

I think it's worth considering if maybe consumer behavior modification is worth putting into a metaresource or Policy-style object instead of re-using the Service for yet another use case.

Reply:
![2022-11-01_08-26-45](https://user-images.githubusercontent.com/623453/199271431-b82d9d9b-868b-49ca-a68b-068b031e3731.png)

In my head this diagram made a lot more sense... but here it is. The curvy lines are the logical flow of requests.

The idea is when you attach a route to a Service (frontend), you are _augmenting_ that frontend. When requests are sent to it, instead, the route rules are applied. In the example above, we hit the frontend from consumer land, so we use the consumer route in place of the service frontend. Then the route sends to the service _backend_, which doesn't have a route associated (since we don't have per-backend routes). So in effect, the consumer route takes precedence.

While its not ideal, I think its the least bad option. See alternatives below.

One thing to note, as well, is this is assuming we are sending to the service _backend_. In theory, we could have `backendRef: ServiceFrontend` and then have it go through both routes. However, we don't have a way to express this in the API and sidecars cannot implement it.

Alternatives:
* Do not have consumer routes: this is a fairly important feature for mesh users, and we still will need to solve it for Gateways at least (which are the ~same problem)
* Merge routes across the two using precedence rules: doesn't really solve any use cases; the routes become highly coupled and small changes to one may result in drastic changes in behavior
* Apply any matching consumer rules, then any matching producer rules. This actually could be viable, but doesn't make sense at all for gateways
* Apply *both* routes. Most implementations cannot do this

Reply:
I think that having the consumer route take precedence over the producer route is fine for now – I think it's a great way to get to something implementable, and I think that's the really critical thing at the moment.

On the other hand, I suspect that the moment we try to use this bit in anger, we'll find that we really, really want some variety of merge behavior. 🙂 It feels like there are a lot of use cases where what the consumer really wants is just to slightly restrict or modify something, not to replace everything wholesale. An example from @howardjohn's PDF jumps out here: perhaps the only thing the consumer needs to do is to rewrite a no-longer-supported HTTP path into its modern equivalent, to let an old client workload keep functioning until it can be replaced. It's easy to imagine a world where the producer route is complex and perhaps changing, and the consumer route really, really doesn't want that burden.

(It's also easy to imagine a world in which, say, the producer specifies a 5-second timeout, and the consumer really wants a 1-second timeout because the client workload has special requirements. In my head, though, that's policy rather than routing, and policy is its own thing entirely.)

So: I agree with @howardjohn that consumer-route-wins is far from ideal, but I'd vote to land this as-is and get to implementing so that we can learn more about what we really need.

Reply:
Seems like consensus is to have "consumer wins" for short term, at least.